### PR TITLE
100% coverage for oauth2client.tools

### PR DESCRIPTION
**Pull request by theacodes**
Towards #212.

### CODE REVIEW

File: tests/test_tools.py [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
-import unittest
 from oauth2client import tools
+from oauth2client.client import OOB_CALLBACK_URN
+from oauth2client.client import FlowExchangeError
+import mock
 from six.moves.urllib import request
+import socket
+import sys
 import threading
+import unittest2
```
Comment (dhermes) [inline tests/test_tools.py:21]: Suggested import order (ordered by stdlib, 3rd party, local):

``` python
import socket
import sys
import threading

import mock
from six.moves.urllib import request
import unittest2

from oauth2client.client import FlowExchangeError
from oauth2client.client import OOB_CALLBACK_URN
from oauth2client import tools
```

File: tests/test_tools.py [hunk 2]
```diff
@@ -27,5 +32,114 @@ def test_ClientRedirectServer(self):
...
 
 
+class _Namespace(object):
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+
+class TestRunFlow(unittest2.TestCase):
+    @mock.patch.object(sys, 'argv',
+        [sys.executable, '--noauth_local_webserver'])
```
Comment (dhermes) [inline tests/test_tools.py]: The executable isn't typically part of `argv`, rather it's the script name.

File: tests/test_tools.py [hunk 3]
```diff
@@ -27,5 +32,114 @@ def test_ClientRedirectServer(self):
...
+    @mock.patch('oauth2client.tools.logging')
+    @mock.patch('oauth2client.tools.input')
+    def test_run_flow_no_webserver(self, input_mock, logging_mock):
+        flow = mock.Mock()
+        storage = mock.Mock()
+        credentials = mock.Mock()
```
Comment (dhermes) [inline tests/test_tools.py]: What's the difference between `mock.Mock` and `mock.MagicMock`?

File: tests/test_tools.py [hunk 4]
```diff
@@ -27,5 +32,114 @@ def test_ClientRedirectServer(self):
...
+        returned_credentials = tools.run_flow(flow, storage)
+
+        self.assertEqual(credentials, returned_credentials)
+        self.assertEqual(flow.redirect_uri, OOB_CALLBACK_URN)
+        flow.step2_exchange.assert_called_once_with('auth_code', http=None)
+        storage.put.assert_called_once_with(credentials)
+        credentials.set_store.assert_called_once_with(storage)
+
+        # Same thing, but with explicit flags
+        flags = _Namespace(noauth_local_webserver=True, logging_level='INFO')
```
Comment (dhermes) [inline tests/test_tools.py]: Fancy class not needed?

``` python
>>> import argparse
>>> argparse.Namespace(noauth_local_webserver=True, logging_level='INFO')
Namespace(logging_level='INFO', noauth_local_webserver=True)
```

File: tests/test_tools.py [hunk 5]
```diff
@@ -27,5 +32,114 @@ def test_ClientRedirectServer(self):
...
+        returned_credentials = tools.run_flow(flow, storage, flags=flags)
+        self.assertEqual(credentials, returned_credentials)
```
Comment (dhermes) [inline tests/test_tools.py]: Can you just make this a separate test (i.e. one with default flags and one with custom flags)?

File: tests/test_tools.py [hunk 6]
```diff
@@ -27,5 +32,114 @@ def test_ClientRedirectServer(self):
...
+
+        # Error while exchanging.
+        with self.assertRaises(SystemExit):
+            flow.step2_exchange.side_effect = FlowExchangeError()
+            returned_credentials = tools.run_flow(flow, storage)
```
Comment (dhermes) [inline tests/test_tools.py]: Ditto here on separate test

File: tests/test_tools.py [hunk 7]
```diff
@@ -27,5 +32,114 @@ def test_ClientRedirectServer(self):
...
+        self.assertEqual(flow.redirect_uri, OOB_CALLBACK_URN)
+        flow.step2_exchange.assert_called_once_with('auth_code', http=None)
+        storage.put.assert_called_once_with(credentials)
+        credentials.set_store.assert_called_once_with(storage)
+        self.assertTrue(server_ctor_mock.called)
+
+
+class TestMessageIfMissing(unittest2.TestCase):
+    def test_message_if_missing(self):
+        self.assertIn('somefile.txt', tools.message_if_missing('somefile.txt'))
```
Comment (dhermes) [inline tests/test_tools.py:195]: Ha. What in the world is `message_if_missing` for?

**Comment by dhermes** [comment]
Thanks a lot for doing this!

File: tests/test_tools.py (hunk 7 above)
Comment (theacodes) [inline tests/test_tools.py:195]: ¯\_(ツ)_/¯ I really have no clue, maybe for downstream users?

Comment (dhermes) [inline tests/test_tools.py:195]: Which stream?

File: tests/test_tools.py (hunk 4 above)
Comment (theacodes) [inline tests/test_tools.py]: Yeah I suppose. py2.6 will throw a fit, but I can skip the test as it none of this works on 2.6 anyway.

File: tests/test_tools.py (hunk 3 above)
Comment (theacodes) [inline tests/test_tools.py]: MagicMock also mocks the magic (`__x__`) methods, typically you don't need to use it unless you actually need those methods.

File: tests/test_tools.py (hunk 2 above)
Comment (theacodes) [inline tests/test_tools.py]: It really doesn't matter in this case, but I can change it to 'ignored' if you want.

File: tests/test_tools.py (hunk 3 above)
Comment (dhermes) [inline tests/test_tools.py]: TIL. w00t! That's a DERP once you hear it for sure.

File: tests/test_tools.py (hunk 2 above)
Comment (dhermes) [inline tests/test_tools.py]: I want

File: tests/test_tools.py (hunk 5 above)
Comment (theacodes) [inline tests/test_tools.py]: will do, it's just a lot of repeated code. I'll see if I can't make it somewhat readable.

File: tests/test_tools.py (hunk 4 above)
Comment (dhermes) [inline tests/test_tools.py]: Yeah. Burn Py2.6 to the ground

**Comment by jonparrott** [comment]
No worries. This module is particularly crufty, so it's important to have these tests before we try to clean it up.

**Comment by dhermes** [comment]
:+1: :+1: :+1: Yaaaaas

File: tests/test_tools.py (hunk 4 above)
Comment (theacodes) [inline tests/test_tools.py]: I remember why I do this- Namespace isn't publicly documented member of argparse. I don't mind either way, so I'll go with using `argparse.Namespace`.

Comment (dhermes) [inline tests/test_tools.py]: :+1:

File: tests/test_tools.py (hunk 7 above)
Comment (theacodes) [inline tests/test_tools.py:195]: samples that use `tools.run_flow`, I guess? I'll see if I can find actual users at another time.

**Comment by jonparrott** [comment]
All comments addressed. Many more tests, but I tried to keep some of the repetition to a minimum.

File: tests/test_tools.py [hunk 8]
```diff
@@ -1,12 +1,22 @@
-"""Unit tests for oauth2client.tools."""
+import socket
```
Comment (dhermes) [inline tests/test_tools.py:15]: I didn't realize there was no license header. Can you add it?

File: tests/test_tools.py [hunk 9]
```diff
@@ -1,12 +1,22 @@
...
+import unittest2
+
+from oauth2client.client import FlowExchangeError
+from oauth2client.client import OOB_CALLBACK_URN
+from oauth2client import tools
+
+try:
+    import argparse
+except ImportError:  # pragma: NO COVER
+    raise unittest2.SkipTest('argparase unavailable.')
```
Comment (dhermes) [inline tests/test_tools.py:30]: TIL!

**Comment by dhermes** [comment]
LGTM (pending license)

Not sure if @nathanielmanistaatgoogle has a preference for module docstrings for tests. It's irrelevant to me.

**Comment by nathanielmanistaatgoogle** [comment]
Looks good (pending license header).

File: tests/test_tools.py (hunk 9 above)
Comment (theacodes) [inline tests/test_tools.py:30]: Only works because nose.

**Comment by jonparrott** [comment]
License header added. Will merge once travis is green.
